src/NUnitEngine/nunit.engine/Services/TestAgency.cs

```diff
@@ -172,47 +190,6 @@ public void StartService()
                 throw;
             }
         }
-
-        private void Release(Guid agentId, ITestAgent agent)
-        {
-            if (_agentStore.IsAgentProcessActive(agentId, out var process))
-            {
-                try
-                {
-                    log.Debug("Stopping remote agent");
-                    agent.Stop();
-                }
-                catch (SocketException ex)
-                {
-                    int? exitCode;
-                    try
-                    {
-                        exitCode = process.ExitCode;
-                    }
-                    catch (NotSupportedException)
-                    {
-                        exitCode = null;
-                    }
-
-                    if (exitCode == 0)
-                    {
-                        log.Warning("Agent connection was forcibly closed. Exit code was 0, so agent shutdown OK");
-                    }
-                    else
-                    {
-                        var stopError = $"Agent connection was forcibly closed. Exit code was {exitCode?.ToString() ?? "unknown"}. {Environment.NewLine}{ExceptionHelper.BuildMessageAndStackTrace(ex)}";
-                        log.Error(stopError);
-                        throw new NUnitEngineUnloadException(stopError, ex);
-                    }
-                }
-                catch (Exception ex)
-                {
-                    var stopError = "Failed to stop the remote agent." + Environment.NewLine + ExceptionHelper.BuildMessageAndStackTrace(ex);
-                    log.Error(stopError);
-                    throw new NUnitEngineUnloadException(stopError, ex);
-                }
-            }
-        }
```

Comment:
I'm in favor of keeping communication separate from the agency itself and keeping both remoting and TCP - at least for now. I believe I removed this because it was never called, but I'll check.

Reply:
This one is complicated. I made all the other fixes in the most recent commit and I'll look this over more closely. I think it needs to exist in some form for all transports. So... one more commit to come.

Reply:
One key problem: removing this makes no tests fail. That may not mean that we don't need it,  just that it was never tested.

Reply:
Heh - I think testing of all of our remote-communications stuff is a little 'sparse'! Room for improvement there!

Reply:
@ChrisMaddock 

The basic problem is that an agent - particularly one running in a process - can terminate in many different ways, e.g.:
* It is told to terminate and shuts down normally
* It is told to terminate and fails to do it cleanly
* It encounters an unexpected exception and dies
* It keeps running but the communications link is lost

It's definitely a can of worms and we have been dealing with it piecemeal. I want to do a separate PR around the creation and stopping of agents, all as a part of implementing #923. That should include documenting how agents work modifying the interfaces and probably adding one or more new statuses to AgentStore (e.g. agent has signalled it is stopping but the process has not yet terminated).

In the latest commit, I merely restored the method and then commented it out... I don't want to forget about it.

I think I'll probably do the planned second PR (adding TCP transport) first and then do the above separately.

If that's OK with you, I'll merge.